cpp/src/fil/infer.cu

```diff
@@ -236,19 +236,51 @@ struct tree_aggregator_t {
     : tmp_storage(finalize_workspace) {}
 
   __device__ __forceinline__ void accumulate(
-    vec<NITEMS, float> single_tree_prediction, int tree, int num_rows) {
+    vec<NITEMS, float> single_tree_prediction, int tree) {
     acc += single_tree_prediction;
   }
 
-  __device__ __forceinline__ void finalize(float* out, int num_rows,
+  __device__ __forceinline__ void finalize(float* block_out, int block_num_rows,
                                            int output_stride,
-                                           output_t transform, int num_trees) {
+                                           output_t transform, int num_trees,
+                                           int log2_threads_per_tree) {
+    auto per_thread = (vec<NITEMS, float>*)tmp_storage;
+    if (log2_threads_per_tree < 5) {  // >1 tree/warp
```

Comment:
This special case speeds up inference by ~30% across `log2_threads_per_tree` values. Models being similar to ones we used for GTC. This is also faster than CUB for 1 thread/tree, at least on Volta.
Perhaps that's because we're heavy on L1/shmem wavefronts, and CUB assumes all blocks do the same thing. That is, over-relying on shuffles frees some wavefronts for blocks traversing trees, even if it isn't a generally optimal strategy